# Information Retrieval System - (Four Stories)

#### 1.  AIR CONDITIONING
#### 2. THE FABLE OF THE ANT AND THE CRICKET
#### 3. THE SEVEN OLD SAMURAI
#### 4. THE THREE WISHES

In [1]:
import math
import pandas as pd

### Reading Story 1 "AIR CONDITIONING"

In [1]:
file1 = open("aircon.txt","r") 

# read all lines at once
text1 = file1.read()

# close the file
file1.close()

#### Processing "Air Conditioning"

In [2]:
# split into words
from nltk.tokenize import word_tokenize
tokens = word_tokenize(text1)
# convert to lower case
tokens = [w.lower() for w in tokens]

In [3]:
# remove punctuation from each word
import string
table = str.maketrans('', '', string.punctuation)
stripped = [w.translate(table) for w in tokens]

In [4]:
# remove remaining tokens that are not alphabetic
words = [word for word in stripped if word.isalpha()]

In [5]:
# filter out stop words
from nltk.corpus import stopwords
stop_words = set(stopwords.words('english'))
data1 = [w for w in words if not w in stop_words]
print(words[:100])

['air', 'conditioning', 'my', 'telephone', 'receiver', 'slams', 'down', 'on', 'its', 'cradle', 'im', 'upset', 'i', 'am', 'soaked', 'to', 'the', 'skin', 'sweat', 'runs', 'from', 'my', 'brow', 'the', 'air', 'conditioner', 'that', 'i', 'so', 'naively', 'entrusted', 'to', 'the', 'yellow', 'pages', 'repair', 'shop', 'is', 'delayed', 'another', 'two', 'weeks', 'i', 'could', 'have', 'it', 'back', 'tomorrow', 'i', 'was', 'told', 'if', 'i', 'happen', 'to', 'have', 'a', 'compressor', 'relief', 'control', 'valve', 'sensor', 'assembly', 'part', 'number', 'in', 'my', 'pocket', 'the', 'repairman', 'is', 'a', 'funny', 'fellow', 'very', 'funny', 'its', 'a', 'bit', 'stuffy', 'in', 'here', 'my', 'secretary', 'says', 'in', 'an', 'attempt', 'to', 'explain', 'her', 'entering', 'my', 'office', 'this', 'is', 'obvious', 'of', 'course', 'as']


### Reading Story 2 "The Fable of the Ant and the Cricket"

In [6]:
file2 = open("antcrick.txt","r") 

# read all lines at once
text2 = file2.read()

# close the file
file2.close()


#### Processing "The Fable of the Ant and the Cricket""

In [7]:
# split into words
from nltk.tokenize import word_tokenize
tokens = word_tokenize(text2)
# convert to lower case
tokens = [w.lower() for w in tokens]
# remove punctuation from each word
import string
table = str.maketrans('', '', string.punctuation)
stripped = [w.translate(table) for w in tokens]
# remove remaining tokens that are not alphabetic
words = [word for word in stripped if word.isalpha()]
# filter out stop words
from nltk.corpus import stopwords
stop_words = set(stopwords.words('english'))
data2 = [w for w in words if not w in stop_words]
print(words[:100])

['the', 'fable', 'of', 'the', 'ant', 'and', 'the', 'cricket', 'once', 'upon', 'a', 'time', 'one', 'hot', 'summer', 'a', 'cricket', 'sang', 'cheerfully', 'on', 'the', 'branch', 'of', 'a', 'tree', 'while', 'down', 'below', 'a', 'long', 'line', 'of', 'ants', 'struggled', 'damely', 'under', 'the', 'weight', 'of', 'their', 'load', 'of', 'grains', 'and', 'between', 'one', 'song', 'and', 'the', 'next', 'the', 'cricket', 'spoke', 'to', 'the', 'ants', 'why', 'are', 'you', 'working', 'so', 'hard', 'come', 'into', 'the', 'shade', 'away', 'from', 'the', 'sun', 'and', 'sing', 'a', 'song', 'with', 'me', 'but', 'the', 'tireless', 'ants', 'went', 'on', 'with', 'the', 'work', 'we', 'ca', 'nt', 'do', 'that', 'they', 'said', 'we', 'must', 'store', 'away', 'food', 'for', 'the', 'winter']


### Reading Story 3 "THE SEVEN OLD SAMURAI"

In [8]:
file3 = open("7oldsamr.txt","r") 

# read all lines at once
text3 = file3.read()

# close the file
file3.close()

#### Processing "THE SEVEN OLD SAMURAI""

In [9]:
# split into words
from nltk.tokenize import word_tokenize
tokens = word_tokenize(text3)
# convert to lower case
tokens = [w.lower() for w in tokens]
# remove punctuation from each word
import string
table = str.maketrans('', '', string.punctuation)
stripped = [w.translate(table) for w in tokens]
# remove remaining tokens that are not alphabetic
words = [word for word in stripped if word.isalpha()]
# filter out stop words
from nltk.corpus import stopwords
stop_words = set(stopwords.words('english'))
data3 = [w for w in words if not w in stop_words]
print(words[:100])

['the', 'seven', 'old', 'samurai', 'once', 'upon', 'a', 'time', 'in', 'far', 'off', 'japan', 'a', 'band', 'of', 'fierce', 'robbers', 'had', 'their', 'hiding', 'place', 'on', 'top', 'of', 'a', 'mountain', 'almost', 'always', 'covered', 'with', 'grey', 'clouds', 'windswept', 'and', 'battered', 'by', 'storms', 'the', 'robbers', 'lived', 'in', 'a', 'large', 'cave', 'where', 'they', 'had', 'piled', 'their', 'spoils', 'now', 'and', 'again', 'they', 'went', 'down', 'the', 'mountain', 'attacked', 'a', 'village', 'murdered', 'the', 'poor', 'folk', 'they', 'chanced', 'upon', 'stole', 'whatever', 'they', 'could', 'lay', 'hands', 'on', 'and', 'burned', 'it', 'to', 'the', 'ground', 'wherever', 'the', 'robbers', 'passed', 'there', 'was', 'nothing', 'but', 'smoking', 'ruins', 'weeping', 'men', 'and', 'women', 'misery', 'mournlng', 'and', 'desolation', 'the']


### Reading Story 4 "THE THREE WISHES"

In [10]:
file4 = open("3wishes.txt","r") 

# read all lines at once
text4 = file4.read()

# close the file
file4.close()

#### Processing "THE THREE WISHES""


In [11]:
# split into words
from nltk.tokenize import word_tokenize
tokens = word_tokenize(text4)
# convert to lower case
tokens = [w.lower() for w in tokens]
# remove punctuation from each word
import string
table = str.maketrans('', '', string.punctuation)
stripped = [w.translate(table) for w in tokens]
# remove remaining tokens that are not alphabetic
words = [word for word in stripped if word.isalpha()]
# filter out stop words
from nltk.corpus import stopwords
stop_words = set(stopwords.words('english'))
data4 = [w for w in words if not w in stop_words]
print(words[:100])

['the', 'three', 'wishes', 'once', 'upon', 'a', 'time', 'a', 'woodcutter', 'lived', 'happily', 'with', 'his', 'wife', 'in', 'a', 'pretty', 'little', 'log', 'cabin', 'in', 'the', 'middle', 'of', 'a', 'thick', 'forest', 'each', 'morning', 'he', 'set', 'off', 'singing', 'to', 'work', 'and', 'when', 'he', 'came', 'home', 'in', 'the', 'evening', 'a', 'plate', 'of', 'hot', 'steaming', 'soup', 'was', 'always', 'waiting', 'for', 'him', 'one', 'day', 'however', 'he', 'had', 'a', 'strange', 'surprise', 'he', 'came', 'upon', 'a', 'big', 'fir', 'tree', 'with', 'strange', 'open', 'holes', 'on', 'the', 'trunk', 'it', 'looked', 'somehow', 'different', 'from', 'the', 'other', 'trees', 'and', 'just', 'as', 'he', 'was', 'about', 'to', 'chop', 'it', 'down', 'the', 'alarmed', 'face', 'of', 'an', 'elf']


In [12]:
data = [data1, data2, data3, data4]

# Data Structures Implementation

# 1. Inverted Index 

In [13]:
from collections import defaultdict

inv_indx = defaultdict(list)
for idx, text in enumerate(data):
    for word in text:
        inv_indx[word].append(idx)

In [14]:
inv_indx

defaultdict(list,
            {'air': [0, 0, 0, 3],
             'conditioning': [0],
             'telephone': [0],
             'receiver': [0],
             'slams': [0],
             'cradle': [0],
             'im': [0],
             'upset': [0, 3],
             'soaked': [0],
             'skin': [0],
             'sweat': [0, 0],
             'runs': [0],
             'brow': [0],
             'conditioner': [0, 0],
             'naively': [0],
             'entrusted': [0],
             'yellow': [0, 1],
             'pages': [0],
             'repair': [0],
             'shop': [0, 0],
             'delayed': [0],
             'another': [0, 3],
             'two': [0],
             'weeks': [0, 1, 1],
             'could': [0, 1, 2, 2, 3, 3, 3],
             'back': [0],
             'tomorrow': [0],
             'told': [0, 3],
             'happen': [0],
             'compressor': [0],
             'relief': [0],
             'control': [0],
             'valve': [0],
    

In [15]:
def CountFrequency(list): 
      
    # Creating an empty dictionary  
    freq = {} 
    for items in my_list: 
        freq[items] = my_list.count(items) 
        
        
    for key, value in freq.items(): 
        print ("% d : % d"%(key, value))

In [16]:
query1 = "last"
print ("Query: "+ query1)
my_list = inv_indx[query1]
documentoccur = set(my_list)

print ("The given term occurs in the document number:")
print (documentoccur)
print(" ")
print("Occurence of term in document number and their frequency:") 
print(documentoccur)
result = CountFrequency(my_list)

print (" \n")
query2 = "cricket"
print ("Query: "+ query2)
my_list = inv_indx[query2]
documentoccur = set(my_list)

print ("The given term occurs in the document number:")
print (documentoccur)
print(" ")
print("Occurence of term in document number and their frequency:") 
print(documentoccur)
result = CountFrequency(my_list)

Query: last
The given term occurs in the document number:
{0, 1, 2, 3}
 
Occurence of term in document number and their frequency:
{0, 1, 2, 3}
 0 :  1
 1 :  2
 2 :  3
 3 :  1
 

Query: cricket
The given term occurs in the document number:
{1}
 
Occurence of term in document number and their frequency:
{1}
 1 :  10


# 2. Boolean Retrieval

In [17]:
import nltk
import glob
import re
import os
import numpy as np
import sys

In [18]:
all_words = []
files_with_index = {'1:"aircon"', '2:"antcrick"', '3:"7oldsamr"', '4:"3wishes"'}
unique_words_all = set(data1).union(set(data2)).union(set(data3)).union(set(data4))

In [19]:
def finding_all_unique_words_and_freq(words):
    words_unique = []
    word_freq = {}
    for word in words:
        if word not in words_unique:
            words_unique.append(word)
    for word in words_unique:
        word_freq[word] = words.count(word)
    return word_freq
def finding_freq_of_word_in_doc(word,words):
    freq = words.count(word)

In [20]:
class Node:
    def __init__(self ,docId, freq = None):
        self.freq = freq
        self.doc = docId
        self.nextval = None
    
class SlinkedList:
    def __init__(self ,head = None):
        self.head = head

In [21]:
linked_list_data = {}
for word in unique_words_all:
    linked_list_data[word] = SlinkedList()
    linked_list_data[word].head = Node(1,Node)
word_freq_in_doc = {}
idx = 1


In [22]:
for item in data:
    words = item
    words = [word for word in words if len(words)>1]
    words = [word.lower() for word in words]
    word_freq_in_doc = finding_all_unique_words_and_freq(words)
    for word in word_freq_in_doc.keys():
        linked_list = linked_list_data[word].head
        while linked_list.nextval is not None:
            linked_list = linked_list.nextval
        linked_list.nextval = Node(idx ,word_freq_in_doc[word])
    idx = idx + 1

In [26]:
query = input('Enter your query:')
query = word_tokenize(query)
connecting_words = []
cnt = 1
different_words = []
for word in query:
    if word.lower() != "and" and word.lower() != "or" and word.lower() != "not":
        different_words.append(word.lower())
    else:
        connecting_words.append(word.lower())
print(connecting_words)
total_files = len(files_with_index)
zeroes_and_ones = []
zeroes_and_ones_of_all_words = []
for word in (different_words):
    if word.lower() in unique_words_all:
        zeroes_and_ones = [0] * total_files
        linkedlist = linked_list_data[word].head
        print(word)
        while linkedlist.nextval is not None:
            zeroes_and_ones[linkedlist.nextval.doc - 1] = 1
            linkedlist = linkedlist.nextval
        zeroes_and_ones_of_all_words.append(zeroes_and_ones)
    else:
        print(word," not found")
        sys.exit()
print(zeroes_and_ones_of_all_words)
for word in connecting_words:
    word_list1 = zeroes_and_ones_of_all_words[0]
    word_list2 = zeroes_and_ones_of_all_words[1]
    if word == "and":
        bitwise_op = [w1 & w2 for (w1,w2) in zip(word_list1,word_list2)]
        zeroes_and_ones_of_all_words.remove(word_list1)
        zeroes_and_ones_of_all_words.remove(word_list2)
        zeroes_and_ones_of_all_words.insert(0, bitwise_op);
    elif word == "or":
        bitwise_op = [w1 | w2 for (w1,w2) in zip(word_list1,word_list2)]
        zeroes_and_ones_of_all_words.remove(word_list1)
        zeroes_and_ones_of_all_words.remove(word_list2)
        zeroes_and_ones_of_all_words.insert(0, bitwise_op);
    elif word == "not":
        bitwise_op = [not w1 for w1 in word_list2]
        bitwise_op = [int(b == True) for b in bitwise_op]
        zeroes_and_ones_of_all_words.remove(word_list2)
        zeroes_and_ones_of_all_words.remove(word_list1)
        bitwise_op = [w1 & w2 for (w1,w2) in zip(word_list1,bitwise_op)]
zeroes_and_ones_of_all_words.insert(0, bitwise_op);
        
files = []    
print(zeroes_and_ones_of_all_words)
lis = zeroes_and_ones_of_all_words[0]
cnt = 4
for index in lis:
    if index == 1:
        docid = index + 1
        print("Present in Document No:")
        print(docid)
    


Enter your query:local AND supply
['and']
local
supply
[[1, 0, 0, 0], [1, 0, 0, 0]]
[[1, 0, 0, 0], [1, 0, 0, 0]]
Present in Document No:
2


# 3. Jaccard Coefficient

In [129]:
def jaccard_similarity(query, document):
    intersection = set(query).intersection(set(document))
    union = set(query).union(set(document))
    return len(intersection)/len(union)
#https://medium.com/@adriensieg/text-similarities-da019229c894

# Jaccard Coefficient Short Query

In [132]:
query = "Who cares about seven beggars".split()
jaccard_similarity(query, data1)

0.0

In [133]:
query = "Who cares about seven beggars".split()
jaccard_similarity(query, data2)

0.0

In [134]:
query = "Who cares about seven beggars".split()
jaccard_similarity(query, data3)

0.017543859649122806

In [135]:
query = "Who cares about seven beggars".split()
jaccard_similarity(query, data4)

0.0

# Jaccard Coefficient Long Query

In [137]:
query = "An early frost tinged the fields with white and turned the last of the green leaves brown"
jaccard_similarity(query, data1)

0.0

In [138]:
query = "An early frost tinged the fields with white and turned the last of the green leaves brown"
jaccard_similarity(query, data2)

0.0

In [139]:
query = "An early frost tinged the fields with white and turned the last of the green leaves brown"
jaccard_similarity(query, data3)

0.0

In [140]:
query = "An early frost tinged the fields with white and turned the last of the green leaves brown"
jaccard_similarity(query, data4)

0.0

# 4. TF-IDF

In [119]:
import gensim
print(dir(gensim))

['__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '__version__', 'corpora', 'interfaces', 'logger', 'logging', 'matutils', 'models', 'parsing', 'scripts', 'similarities', 'summarization', 'topic_coherence', 'utils']


In [120]:
dictionary = gensim.corpora.Dictionary(data)

In [121]:
print("Number of words in dictionary:",len(dictionary))


Number of words in dictionary: 606


In [122]:
corpus = [dictionary.doc2bow(gen_doc) for gen_doc in data]
#print(corpus)

In [123]:
tf_idf = gensim.models.TfidfModel(corpus)
print(tf_idf)
s = 0
for i in corpus:
    s += len(i)
print(s)

TfidfModel(num_docs=4, num_nnz=694)
694


In [124]:
sims = gensim.similarities.Similarity('C:/Users/Inaam/Desktop/NLP/Similarity/sims',tf_idf[corpus], num_features=len(dictionary))
print(sims)
print(type(sims))

Similarity index with 4 documents in 0 shards (stored under C:/Users/Inaam/Desktop/NLP/Similarity/sims)
<class 'gensim.similarities.docsim.Similarity'>


# TF-IDF Short Query 

In [125]:
query_doc = [w.lower() for w in word_tokenize("Who cares about seven beggars.")]
print(query_doc)
query_doc_bow = dictionary.doc2bow(query_doc)
print(query_doc_bow)
query_doc_tf_idf = tf_idf[query_doc_bow]
print(query_doc_tf_idf)

['who', 'cares', 'about', 'seven', 'beggars', '.']
[(288, 1), (295, 1), (388, 1)]
[(288, 0.5773502691896258), (295, 0.5773502691896258), (388, 0.5773502691896258)]


In [126]:
print('Query result:')
sims[query_doc_tf_idf]

Query result:


array([0.        , 0.        , 0.20997019, 0.        ], dtype=float32)

### TF-IDF predict correctly with score 0.209, the text is from story 3


# TF-IDF Long Query

In [127]:
query_doc = [w.lower() for w in word_tokenize("An early frost tinged the fields with white and turned the last of the green leaves brown")]
print(query_doc)
query_doc_bow = dictionary.doc2bow(query_doc)
print(query_doc_bow)
query_doc_tf_idf = tf_idf[query_doc_bow]
print(query_doc_tf_idf)

['an', 'early', 'frost', 'tinged', 'the', 'fields', 'with', 'white', 'and', 'turned', 'the', 'last', 'of', 'the', 'green', 'leaves', 'brown']
[(68, 1), (164, 1), (177, 1), (187, 1), (191, 1), (198, 1), (206, 1), (252, 1), (258, 1), (268, 1)]
[(164, 0.3333333333333333), (177, 0.3333333333333333), (187, 0.3333333333333333), (191, 0.3333333333333333), (198, 0.3333333333333333), (206, 0.3333333333333333), (252, 0.3333333333333333), (258, 0.3333333333333333), (268, 0.3333333333333333)]


In [128]:
print('Query result:')
sims[query_doc_tf_idf]

Query result:


array([0.        , 0.22125918, 0.        , 0.        ], dtype=float32)

### TF-IDF predict correctly with score 0.221, the text is from story 2

# 5. Cosine Similarity

In [105]:
query = "An early frost tinged the fields with white and turned the last of the green leaves brown".split()

In [106]:
wordSet = set(query).union(set(data1)).union(set(data2)).union(set(data3)).union(set(data4))

In [107]:
queryDict = dict.fromkeys(wordSet, 0)
wordDictA = dict.fromkeys(wordSet, 0) 
wordDictB = dict.fromkeys(wordSet, 0)
wordDictC = dict.fromkeys(wordSet, 0)
wordDictD = dict.fromkeys(wordSet, 0)

In [108]:
for word in query:
    queryDict[word]+=1
    
for word in data1:
    wordDictA[word]+=1
    
for word in data2:
    wordDictB[word]+=1
    
for word in data3:
    wordDictC[word]+=1
    
for word in data4:
    wordDictD[word]+=1

In [109]:
import pandas as pd
pd.DataFrame([queryDict, wordDictA, wordDictB, wordDictC, wordDictD])

An  aback  active  adjusts  afraid  agree  agreed  ah  air  alarmed  ...  \
0   1      0       0        0       0      0       0   0    0        0  ...   
1   0      0       1        1       0      0       0   0    3        0  ...   
2   0      0       0        0       0      0       0   1    0        0  ...   
3   0      0       0        0       0      0       0   0    0        0  ...   
4   0      1       0        0       1      1       1   0    1        1  ...   

   work  worked  working  worn  worried  would  years  yellow  yes  young  
0     0       0        0     0        0      0      0       0    0      0  
1     0       0        0     0        0      0      0       1    0      0  
2     2       0        1     0        0      0      0       1    1      0  
3     0       0        0     1        1      1      1       0    0      1  
4     1       1        0     0        0      4      0       0    2      0  

[5 rows x 611 columns]

In [110]:
def computeTF(wordDict, bow):
    tfDict = {}
    bowCount = len(bow)
    for word, count in wordDict.items():
        tfDict[word] = count/float(bowCount)
    return tfDict

In [111]:
tfquery = computeTF(queryDict, query)
tfBowA = computeTF(wordDictA, data1)
tfBowB = computeTF(wordDictB, data2)
tfBowC = computeTF(wordDictC, data3)
tfBowD = computeTF(wordDictD, data4)

In [112]:
def computeIDF(docList):
    import math
    idfDict = {}
    N = len(docList)
    
    idfDict = dict.fromkeys(docList[0].keys(), 0)
    for doc in docList:
        for word, val in doc.items():
            if val > 0:
                idfDict[word] += 1
    
    for word, val in idfDict.items():
        idfDict[word] = math.log10(N / float(val))
        
    return idfDict

In [113]:
idfs = computeIDF([queryDict, wordDictA, wordDictB, wordDictC, wordDictD])


In [114]:
def computeTFIDF(tfBow, idfs):
    tfidf = {}
    for word, val in tfBow.items():
        tfidf[word] = val*idfs[word]
    return tfidf

In [115]:
tfidfquery = computeTFIDF(tfquery, idfs)
tfidfBowA = computeTFIDF(tfBowA, idfs)
tfidfBowB = computeTFIDF(tfBowB, idfs)
tfidfBowC = computeTFIDF(tfBowC, idfs)
tfidfBowD = computeTFIDF(tfBowD, idfs)

In [116]:
import pandas as pd
tfidfscore = pd.DataFrame([tfidfquery, tfidfBowA, tfidfBowB, tfidfBowC, tfidfBowD])

In [117]:
tfidfscore

An     aback   active  adjusts    afraid     agree    agreed  \
0  0.041116  0.000000  0.00000  0.00000  0.000000  0.000000  0.000000   
1  0.000000  0.000000  0.00404  0.00404  0.000000  0.000000  0.000000   
2  0.000000  0.000000  0.00000  0.00000  0.000000  0.000000  0.000000   
3  0.000000  0.000000  0.00000  0.00000  0.000000  0.000000  0.000000   
4  0.000000  0.001952  0.00000  0.00000  0.001952  0.001952  0.001952   

         ah       air   alarmed  ...      work    worked   working      worn  \
0  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000  0.000000   
1  0.000000  0.006901  0.000000  ...  0.000000  0.000000  0.000000  0.000000   
2  0.003738  0.000000  0.000000  ...  0.004256  0.000000  0.003738  0.000000   
3  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000  0.002853   
4  0.000000  0.001112  0.001952  ...  0.001112  0.001952  0.000000  0.000000   

    worried     would     years    yellow       yes     young  
0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  
1  0.000000  0.000000  0.000000  0.002300  0.000000  0.000000  
2  0.000000  0.000000  0.000000  0.002128  0.002128  0.000000  
3  0.002853  0.001624  0.002853  0.000000  0.000000  0.002853  
4  0.000000  0.004446  0.000000  0.000000  0.002223  0.000000  

[5 rows x 611 columns]

# Cosine Similarity for Short Query

In [ ]:
shortquery = "who cares about seven beggars"

In [104]:
from sklearn.metrics.pairwise import cosine_similarity
cosine_similarity(tfidfscore[0:1], tfidfscore)

array([[1.        , 0.        , 0.        , 0.06972064, 0.        ]])

## The given text is from the third story "THE SEVEN OLD SAMURAI" which is correctly predicted with a similarity score of 0.069

# Cosine Similarity for Long Query

In [32]:
longquery = "An early frost tinged the fields with white and turned the last of the green leaves brown"

In [118]:
from sklearn.metrics.pairwise import cosine_similarity
cosine_similarity(tfidfscore[0:1], tfidfscore)

array([[1.        , 0.        , 0.05434862, 0.        , 0.        ]])

## "The Fable of the Ant and the Cricket" is the second story which is correctly predicted by Cosine Similarity with a similarity score = 0.0543